README.md
# fastbill [![Build Status](https://travis-ci.org/konexmedia/fastbill.svg?branch=master)](https://travis-ci.org/konexmedia/fastbill)

A client implementation for communicating with the [FastBill](http://fastbill.com) [API](http://www.fastbill.com/api).

## Installation

    npm install --save fastbill

## Quick Start

Using this client is easy. All you need is an FastBill account and your personal API key. You don't have to register the key manually, it is there. Check your settings page (Settings -> Account).

```javascript
'use strict';

var FastBill = require('fastbill');

var api = FastBill.instantiate({
    email: 'your.email@host.tld',
    apikey: 'your-long-api-key'
});
```

You will have access to the API services after instantiating the communication object.

## Services

FastBill provides several services for interacting with the data in your account. This module provides a clean abstraction from interacting with the API in a raw fashion. The following part describes how to use the different parts of this client.

 - [Customer](#customer)

### Customer

Methods for accessing, creating, updating and deleting your customers.

 - [get](#getoptions-callback)
 - [create](#createcustomer-callback)
 - [update](#updateid-modification-callback)
 - [delete](#deleteid-callback)

#### get(options, callback)

Requests the details from one or more customers. When no filter is set, it returns all customers ([API doc](http://www.fastbill.com/api/en/customer.html#customer.get))

 - `options` An object for controlling the result set
    - `limit` Limits the result set
    - `offset` Result set offset
    - `filter` Filter params for restricting the result set
 - `callback` Error-first style callback

##### Usage example

```javascript
api.consumer.get(function (err, consumers) {
    if (err) {
        return console.error(err);
    }
    
    console.log(consumers);
});
```

#### create(customer, callback)

Creates a new customer ([API doc](http://www.fastbill.com/api/en/customer.html#customer.create))

 - `customer` An object which describes the customer that should be created. See the ([API doc](http://www.fastbill.com/api/en/customer.html#customer.create)) in order to learn the attributes that should be used.
 - `callback` Error-first style callback

##### Usage example

```javascript
var customer = {
    CUSTOMER_TYPE: 'consumer',
    LAST_NAME: 'König',
    COUNTRY_CODE: 'DE',
    PAYMENT_TYPE: 1,
    BANK_NAME: 'Postbank',
    BANK_CODE: 1000000,
    BANK_ACCOUNT_NUMBER: 9000000,
    BANK_ACCOUNT_OWNER: 'König',
    BANK_ACCOUNT_MANDATE_REFERENCE: 'DEPBXXXXXXX'
};

api.consumer.create(customer, function (err, id) {
    if (err) {
        return console.error(err);
    }
    
    console.log('Created customer with id %s', id);
});
```

#### update(id, modification, callback)

Updates the information of a customer ([API doc](http://www.fastbill.com/api/en/customer.html#customer.update)).

- `id` The id of the customer which should be altered.
- `modification` The changed customer properties
- `callback` Error-first style callback

##### Usage example

```javascript
var modification = {
    FIRST_NAME: 'André'
};

api.customer.update(1, modification, function (err) {
    if (err) {
        return console.error(err);
    }
    
    console.log('Updated customer');
});
```

#### delete(id, callback)

Deletes a customer ([API doc](http://www.fastbill.com/api/en/customer.html#customer.delete)).

- `id` The id of the customer which should be deleted.
- `callback` Error-first style callback

##### Usage example

```javascript
api.customer.delete(1, function (err) {
    if (err) {
        return console.error(err);
    }
    
    console.log('Deleted customer.');
});
```
## License

The MIT License (MIT) Copyright (c) 2014 konexmedia

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

HISTORY.md
# UNRELEASED

  * Parameter checking: Using new module for checking parameters up front.
  * Browser support: The API abstraction can be used within the browser now.

# 0.2.1 (2014-08-03)

  * Removed .zedstate file

# 0.2.0 (2014-08-03)

  * Support for all `customer` services (`get`, `create`, `update`, `delete`).

# 0.1.0 (2014-07-31)

  * Initial release


lib/customer.js
/*
 * fastbill
 *
 * Copyright(c) 2014 konexmedia <info@konexmedia.com>
 * MIT Licensed
 *
 */

/**
 * @author André König <andre.koenig@konexmedia.com>
 *
 */

'use strict';

var assert = require('assert');
var util = require('util');

var BaseBroker = require('./base');

exports.instantiate = function instantiate (credentials) {
    return new CustomerBroker(credentials);
};

/**
 * The customer broker which abstracts from the
 * customer services of the FastBill API.
 * 
 * @param {object} credentials The email and api key tupel.
 * 
 */
function CustomerBroker (credentials) {
    this.$namespace = 'customer.';

    BaseBroker.call(this, credentials);
}

util.inherits(CustomerBroker, BaseBroker);

/**
 * Requests the details from one or more customers.
 * When no filter is set, it returns all customers
 * 
 * If no options has been passed, the method will return all
 * available customers
 * 
 * See: http://www.fastbill.com/api/en/customer.html#customer.get
 * 
 * Possible options:
 * 
 *     limit: Limits the result set
 *     offset: ResultSet offset
 *     filter: Filter parameter for restricting the result set
 * 
 * @param {object} options Possible request options (see above)
 * @param {function} callback Error-first callback (err, customers)
 * 
 */
CustomerBroker.prototype.get = function get (options, callback) {

    function onResult (err, resultset) {
        if (err) {
            return callback(err);
        }

        callback(null, resultset.CUSTOMERS);
    }

    if ('function' === typeof options) {
        callback = options;
        options = undefined;
    }

    options = options || {};

    assert.equal(typeof (options), 'object', 'argument "options" should be an object.');
    assert.equal(util.isArray(options), false, 'argument "options" should be an object.');
    assert.equal(typeof (callback), 'function');

    this.$request({
        service: this.$namespace + 'get',
        filter: options.filter,
        limit: options.limit,
        offset: options.offset
    }, onResult);
};

/**
 * Creates a new customer
 * 
 * The customer id of the newly created customer will be passed
 * to the callback function.
 * 
 * See: http://www.fastbill.com/api/en/customer.html#customer.create
 *
 * Usage example:
 * 
 *     var customer = {
 *         CUSTOMER_TYPE: 'consumer',
 *         LAST_NAME: 'König',
 *         COUNTRY_CODE: 'DE',
 *         PAYMENT_TYPE: 1,
 *         BANK_NAME: 'Postbank',
 *         BANK_CODE: 1000000,
 *         BANK_ACCOUNT_NUMBER: 9000000,
 *         BANK_ACCOUNT_OWNER: 'König',
 *         BANK_ACCOUNT_MANDATE_REFERENCE: 'DEPBXXXXXXXXXXXXXXXXXXXXXXX'
 *     };
 *
 *     fastbill.customer.create(customer, function (err, id) {
 *         if (err) {
 *             return console.error(err);
 *         }
 *
 *         console.log(id);
 *     });
 * 
 * @param {object} customer The customer that should be created.
 * @param {function} callback Error-first callback (err, customerId)
 * 
 */
CustomerBroker.prototype.create = function create (customer, callback) {

    function onResult (err, results) {
        if (err) {
            return callback(err);
        }

        callback(null, results.CUSTOMER_ID);
    }

    if ('function' === typeof customer) {
        callback = customer;
        customer = undefined;
    }

    customer = customer || {};

    this.$request({
        service: this.$namespace + 'create',
        data: customer
    }, onResult);
};

/**
 * Updates the information of a customer.
 * 
 * See: http://www.fastbill.com/api/en/customer.html#customer.update
 * 
 * Usage example:
 * 
 *     var modification = {
 *         FIRST_NAME: 'André'
 *     };
 *
 *     fastbill.customer.update(1, modification, function (err) {
 *         if (err) {
 *             return console.error(err);
 *         }
 *     });
 *
 *
 * @param {number} id The id of the customer that should be updated.
 * @param {object} modification The modifications.
 * @param {function} callback Error-first callback (err)
 *
 */
CustomerBroker.prototype.update = function update (id, modification, callback) {
    
    function onResult (err) {
        if (err) {
            return callback(err);
        }

        callback(null);
    }

    assert.equal(typeof (id), 'number', 'argument \'id\' must be a number');
    assert.equal(typeof (modification), 'object', 'argument \'modification\' must be an object');
    assert.equal(typeof (callback), 'function');

    modification.CUSTOMER_ID = id;
    
    this.$request({
        service: this.$namespace + 'update',
        data: modification
    }, onResult);
};

/**
 * Deletes a customer.
 * 
 * See: http://www.fastbill.com/api/en/customer.html#customer.delete
 * 
 * Usage example:
 *
 *     fastbill.customer.delete(1, function (err) {
 *         if (err) {
 *             return console.error(err);
 *         }
 *     });
 *
 * @param {number} id The id of the customer that should be deleted.
 * @param {function} callback Error-first callback (err)
 * 
 */
CustomerBroker.prototype.delete = function del (id, callback) {
    
    function onResult (err) {
        if (err) {
            return callback(err);
        }
    
        callback(null);
    }

    if ('function' === typeof id) {
        callback = id;
        id = undefined;
    }
    
    this.$request({
        service: this.$namespace + 'delete',
        data: {CUSTOMER_ID: id}
    }, onResult);
};